driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -2282,6 +2289,234 @@ public void run() {
+                    return new SchemaRefreshRequest(keyspace, null);
+                if(!this.table.equals(that.table))
+                    return new SchemaRefreshRequest(keyspace, null);
+                return this;
+            }
+
+            @Override
+            public String toString() {
+                if(this.keyspace == null)
+                    return "Refresh ALL";
+                if(this.table == null)
+                    return "Refresh keyspace " + keyspace;
+                return "Refresh keyspace " + keyspace + " table " + table;
+            }
+        }
+
+        private class SchemaRefreshRequestDeliveryCallback implements EventDebouncer.DeliveryCallback<SchemaRefreshRequest> {
+
+            @Override
+            public ListenableFuture<?> deliver(final List<SchemaRefreshRequest> events) {
+                return executor.submit(new ExceptionCatchingRunnable() {
+                    @Override
+                    public void runMayThrow() throws InterruptedException, ExecutionException {
+                        SchemaRefreshRequest coalesced = null;
+                        for (SchemaRefreshRequest request : events) {
+                            coalesced = coalesced == null ? request : coalesced.coalesce(request);
+                        }
+                        assert coalesced != null;
+                        logger.trace("Coalesced schema refresh request: {}", coalesced);
+                        controlConnection.refreshSchema(coalesced.keyspace, coalesced.table);
+                    }
+                });
+            }
+
+        }
+
+        private class NodeRefreshRequest {
+
+            private final Host host;
+
+            private final HostEvent eventType;
+
+            private NodeRefreshRequest(Host host, HostEvent eventType) {
+                this.host = host;
+                this.eventType = eventType;
+            }
+
+            @Override
+            public String toString() {
+                return host + " " + eventType;
+            }
+
+        }
+
+        private class NodeRefreshRequestDeliveryCallback implements EventDebouncer.DeliveryCallback<NodeRefreshRequest> {
+
+            @Override
+            public ListenableFuture<?> deliver(List<NodeRefreshRequest> events) {
+                Map<Host, HostEvent> hosts = new HashMap<Host, HostEvent>();
+                // only keep the last event for each host
+                for (NodeRefreshRequest req : events) {
+                    hosts.put(req.host, req.eventType);
+                }
+                List<ListenableFuture<?>> futures = new ArrayList<ListenableFuture<?>>(hosts.size());
+                for (final Entry<Host, HostEvent> entry : hosts.entrySet()) {
+                    final Host host = entry.getKey();
+                    HostEvent eventType = entry.getValue();
+                    switch (eventType) {
+                        case UP:
+                            schedule(hostUp(host));
```

Comment:
> Shouldn't hostUp be executed right away?

According to [CASSANDRA-8236](https://issues.apache.org/jira/browse/CASSANDRA-8236), the problem exists for both UP and ADDED notifications, so not delaying onUp was most likely an oversight (I also discussed it with Sylvain).

> Don't we also have a problem with this sequence of events: onAdd onUp

I wasn't aware that this sequence is possible. For me it was only ADDED when a node gets added. onAdd sets the host up and creates the pools, so there is no need to execute an onUp after.
Anyway, if we process an UP notification for a host that is not known yet, it's turned into an ADDED notification (see `if (hostUp == null) {...` in `Cluster.handle`), so things should work out well. I'll try your test.